Move refactor: moves/items that ignore immunity/effectiveness

ascriptmaster opened the pull request:
Frankly, this is such a huge refactor that I won't feel safe until like, @Zarel, @Marty-D and 2 other people [EDIT: with write access] have independently verified that this is a good idea and will work. It seems to work ostensibly, and a quick `grep` seems to point towards all `affectedByImmunities` properties from being removed, but I still need more opinions just to make sure. (15 files are being changed here!)

SolarisFox commented on data/moves.js:
```diff
@@ -14119,17 +14119,13 @@ exports.BattleMovedex = {
 			target.negateImmunity['Ground'] = 'IgnoreEffectiveness';
 		},
 		volatileStatus: 'smackdown',
+		ignoreImmunity: {'Ground': true},
```
This change in particular seems like a good idea as it both makes the information about the move easier to access and also easier to read and understand. While Thousand Arrows might currently be a special case move, I would not be surprised if more moves that could use this notation will appear in the future.

SolarisFox commented on data/scripts.js:
```diff
@@ -209,14 +209,15 @@ exports.BattleScripts = {
+			return false;
 		}
 
+
```
Added an extra line (mistake?)

Zarel commented on tools.js:
```diff
@@ -311,6 +311,7 @@ module.exports = (function () {
 				else move.gen = 0;
 			}
 			if (!move.priority) move.priority = 0;
+			if (!move.ignoreImmunity) move.ignoreImmunity = (move.category === 'Status' ? true : {});
```
I'm not exactly a huge fan of instantiating a new Object every time we want to represent "this move doesn't ignore immunities". Why not `false`?

ascriptmaster replied:
It messes up Scrappy, I guess? But I guess Scrappy can initialize the object itself instead (this will means I can probably take it out of the `Tools#getMove`)

Zarel replied:
No reason for `Tools#getMove` _not_ to initialize it.

Zarel commented on mods/gen1/statuses.js:
```diff
@@ -310,8 +310,8 @@ exports.BattleStatuses = {
 
-				if (typeof posData.moveData.affectedByImmunities === 'undefined') {
-					posData.moveData.affectedByImmunities = true;
+				if (typeof posData.moveData.ignoreImmunity === 'undefined') {
```
`typeof ... === 'undefined'` should probably be rewritten `... === undefined` or perhaps `!(... in ...)`.

I wrote that that way back when I didn't really understand JavaScript.

ascriptmaster commented:
I think I still need `Tools#getMove` to fill in the `move.ignoreImmunity` because it was being weird without it.

Joimer commented:
I think it would be better if this PR added mocha tests that create battles in which immunities are involved to see if the immunity results are those we expect.

SolarisFox commented:
Because two more cases of ```affectedByImmunities``` were added in #1727 this pullreq should probably also change those, preferably taking advantage of now being able to easily return the correct results for Thousand Arrows.

Zarel commented:
I like this so far.

ascriptmaster commented:
@SolarisFox, those parts of the code should be fixed now.

I'll work on tests tomorrow...

ascriptmaster commented:
Just so I can verify in the test, @Marty-D: What happens to Normalize Thunder Wave against Ghosts in-game?

ascriptmaster commented:
@Joimer, I added the tests as you asked. However, it also uncovered a new bug: Thousand Arrows against an Electric/Flying activates Weakness Policy even though it does neutral damage. 

The fix I introduced in the next commit works ostensibly (and partially optimizes by halving the number of `Effectiveness` events that need to be run), but the fact that this is coming up means that I need @Marty-D to check the interaction between Thousand Arrows and part-Flying Pokemon whose other type is weak to Ground, especially if the target Pokemon has:
- Filter/Solid Rock
- Shuca Berry
- Wonder Guard

At any rate, perhaps instead of this fix, I should move the `IgnoreEffectiveness` check out of `Battle#getDamage` and into `BattlePokemon#runEffectiveness`.

Slayer95 commented:
I love cb9fd99

ascriptmaster commented:
Well, it's probably actually an incomplete fix. Especially since it can't account for like Wonder Guard Aerodactyl or something.

How's the first commit tho

Joimer commented:
Nice tests, thanks! This is a very good fix. Can we get this through and then check other items that interact with immunities?

ascriptmaster commented:
All the `typeMod` things seem to work properly (manually tested).

All that remains is the Wonder Guard vs Thousand Arrows fix, but I'm tired so I'll look at it tomorrow

Marty-D commented:
This is great! Sorry for the tardiness.

@ascriptmaster, Normalize Thunder Wave does not affect Ghost types, but Normalize Will-O-Wisp does, etc.

For weakness-related effects, the effects only activate if the attack was actually super effective. So in the Thousand Arrows case where an airborne Flying-type target is hit, none of those effects activate because the damage is always neutral. Because the damage is neutral, Wonder Guard does activate. (Example: Wonder Guard Zapdos holding Iron Ball has Wonder Guard activate when targeted by Earth Power. Without Iron Ball, Zapdos is simply naturally immune, which is checked before Wonder Guard so the Ability doesn't activate.)

Off the top of my head only Delta Stream, Iron Ball held by an otherwise airborne Flying type, and Thousand Arrows against an otherwise airborne Flying type actually change effectiveness in this way. Stuff like Thick Fat, Filter, and type-resist Berries merely change the resulting damage (I suppose "type-resist" is a misnomer then).

ascriptmaster commented:
Since we have the `Effectiveness` event now, perhaps we could use that for Thousand Arrows and Iron Ball, and forego `'IgnoreEffectiveness'` entirely haha

ascriptmaster commented:
Conceptual idea of how it would work, probably:

    onEffectiveness: function (typeMod, type, pokemon, move) {
        if (move.type !== 'Ground' || pokemon.negateImmunity['Ground'] || this.getImmunity('Ground', pokemon)) return;
        return 0;
    }

Probably needs a negative priority since returning 0 will abort the rest of the event handling iirc

Slayer95 commented:
That was my first idea when implementing the Thousand Arrows' peculiarity. However, it would require the move to set two different volatiles at different times, which is undesirable imo, and overall doesn't provide any real benefit.

ascriptmaster commented:
Adding `onEffectiveness` on Thousand Arrows and Iron Ball directly (no volatile needed) also fixes the situation, assuming the above code works. It'll mess up `/eff` a little with Thousand Arrows, though.

ascriptmaster commented:
Fixed Wonder Guard. `onEffectiveness` doesn't work on Iron Ball because it needs `negateImmunity` to allow the Ground move to hit in the first place.

ascriptmaster commented:
So is anyone going to verify that the most recent version is good?

Joimer commented:
It seems to be good, but I'd rather wait for @Zarel to comment on this as it's kind of a huge refactor. Passing the tests should be proving it's doing good.

Zarel commented:
I would be fine with accepting this and deploying it slowly.

ascriptmaster commented:
Whoops, it turns out I broke something anyways! Some move wasn't initializing `ignoreImmunity`, probably as part of how `getDamage` creates mock moves when `move` is passed in as a number. Alternatively, we can have those statements check if `ignoreImmunity` is truthy before doing the rest of the check. I can't fix it right now but if someone else wants to do it then that's probably the solution.

Joimer commented:
Yeah, I've just added myself a thunder wave to ground type test to the tests and I nailed down the place where the undefined gets to (move.category === 'Status') which makes it fail with thunder wave.

It was just this on .getMove:
if (!move.ignoreImmunity) move.ignoreImmunity = (move.category === 'Status');
While the move has ignoreImmunity at "false". It would always override Thunder Wave. Fixing it and updating the server.

ascriptmaster commented:
Oh, right. It should be `if (move.ignoreImmunity === undefined)`. I had forgotten that Zarel had gotten me to set it to `false` if no specific type ignoring was needed, to save processing speed.

At any rate, #1784 should patch the rest up, because confusion might cause Gen 1 games to crash without it, I think.

Joimer commented:
Indeed, I've fixed the problem and added your PR as well. Let's hope nothing else happens now.
